pyomo/core/tests/unit/test_componentuid.py

```diff
@@ -0,0 +1,891 @@
+        self.assertEqual(
+            cuid._cids,
+            (('b',('inf',)), ('x',tuple())) )
+
+        self.assertTrue(cuid.matches(x))
+        self.assertEqual(repr(ComponentUID(x)), ref)
+        self.assertEqual(str(ComponentUID(x)), ref)
+
+        ref = r"b:$inf.x"
+        cuid = ComponentUID(ref)
+        self.assertEqual(
+            cuid._cids,
+            (('b',('inf',)), ('x',tuple())) )
+
+        self.assertTrue(cuid.matches(x))
+        self.assertEqual(ComponentUID(x).get_repr(1), ref)
+        self.assertEqual(str(ComponentUID(x)), r"b['inf'].x")
+
+
+    def test_find_component_deprecated(self):
+        ref = self.m.b[1,'2'].c.a[3]
+        cuid = ComponentUID(ref)
+        DEP_OUT = StringIO()
+        with LoggingIntercept(DEP_OUT, 'pyomo.core'):
+            self.assertTrue(cuid.find_component(self.m) is ref)
+        self.assertIn('ComponentUID.find_component() is deprecated.',
+                      DEP_OUT.getvalue())
+
+    def test_find_explicit_exists(self):
+        ref = self.m.b[1,'2'].c.a[3]
+        cuid = ComponentUID(ref)
+        self.assertTrue(cuid.find_component_on(self.m) is ref)
+
+    def test_find_component_exists_1(self):
+        ref = self.m.b[1,'2'].c.a
+        cuid = ComponentUID(ref)
+        self.assertTrue(cuid.find_component_on(self.m) is ref)
+
+    def test_find_wildcard(self):
+        cuid = ComponentUID('b:1,$2.c.a:*')
+        comp = cuid.find_component_on(self.m)
+        self.assertIs(comp.ctype, Param)
+        cList = list(itervalues(comp))
+        self.assertEqual(len(cList), 3)
+        self.assertEqual(cList, list(self.m.b[1,'2'].c.a[:]))
+
+        cuid = ComponentUID('b[*,*]')
+        comp = cuid.find_component_on(self.m)
+        self.assertIs(comp.ctype, Block)
+        cList = list(itervalues(comp))
+        self.assertEqual(len(cList), 9)
+        self.assertEqual(cList, list(self.m.b.values()))
+
+    def test_find_wildcard_partial_exists(self):
+        # proper Reference: to ComponentData
+        cuid = ComponentUID('b[*,*].c.a[**]')
+        comp = cuid.find_component_on(self.m)
+        self.assertIs(comp.ctype, Param)
+        cList = list(itervalues(comp))
+        self.assertEqual(len(cList), 3)
+        self.assertEqual(cList, list(self.m.b[1,'2'].c.a[:]))
+
+        # improper Reference: to IndexedComponent
+        cuid = ComponentUID('b[*,*].c.a')
+        comp = cuid.find_component_on(self.m)
+        self.assertIs(comp.ctype, IndexedComponent)
+        cList = list(itervalues(comp))
+        self.assertEqual(len(cList), 1)
+        self.assertIs(cList[0], self.m.b[1,'2'].c.a)
+
+    def test_find_wildcard_not_exists(self):
+        cuid = ComponentUID('b[*,*].c.x')
+        self.assertIsNone(cuid.find_component_on(self.m))
+
+    # def test_find_implicit_exists(self):
+    #     cuid = ComponentUID('b:1,2.c.a:3')
+    #     self.assertTrue(cuid.find_component_on(self.m) is
+    #                     self.m.b[1,'2'].c.a[3])
```

Comment:
Why is this test commented out?

Reply:
ComponentUID used to support "fuzzy matching" of component names.  That is, the "name" of a component is ambiguous:
```
m.x = Var(Any, dense=False)
m.x[1].name == "x[1]"  # True!
m.x['1'].name == "x[1]"  # True!
m.x['1'] is not m.x[1]  # True!
```
The old ComponentUID would take strings like `"x[1]"` and attempt to perform "implicit" matching (looking first for `1` and then for `'1'`).  This PR removed that functionality: it was too prone to generating error, plus ComponentUID can now correctly parse quoted indices, so it is not as big of an issue.  A future PR will update `Component.name` so that ambiguous strings (containing spaces or commas, or consisting of parsable floating point numbers) will always be quoted.